sys/trace: add trace module

kaspar030 opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR adds a module that can be very helpful for debugging intra-thread or thread-isr interactions.

From the docs:

```
This module allows recording program flow traces. It is meant for debugging                                                     
in multi-threaded applications or when ISR's are involved.                                                                      
                                                                                                                                
The `trace()` function takes an arbitrary (user chosen) uint32 value.                                                           
Calling the function is safe from anywhere (user code, ISR, ...) and safely                                                     
logs the function call time and user value in a trace buffer.   
                                                                                                                                                                                                 
At any point, `trace_dump()` can be used to print the trace buffer.                                                             
                                                                                                                                 
The buffer has a default size of 512 entries, which can be overridden by                                                        
defining TRACE_BUFSIZE. It can be cleared using `trace_reset()`.                                                                
                                                                                                                                 
Tracing is made thread safe by disabling interrupts for critical sections.                                                      
                                                                                                                                 
It does incur some overhead (at least a function call, getting the current                                                      
time, a pair of enable/disable interrupts and a couple of memory accesses).                                                     
```

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

Waiting for #14204 due to name clash.

<!-- 
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

miri64 commented:
There is some functional overlap between new `trace` and now `backtrace`. Would it be thinkable to serve them both with the same API?

kaspar030 commented:
> There is some functional overlap between new `trace` and now `backtrace`. Would it be thinkable to serve them both with the same API?

Is it? one provides access to the toolchain provided stack backtrace, the other provides an arbitrary (time, value) log.

miri64 commented:
> Is it? one provides access to the toolchain provided stack backtrace, the other provides an arbitrary (time, value) log.

I guess I misread your OP then. Maybe it helps to state the fact, that it is (time, value) log somewhere near the first sentence.

kaspar030 commented:
Should I try to add the callsite to the log? It might be possible using '__builtin_return_address()', but would require the noinline attribute. It could be optional if performance is hurt a lot.
@aabadie I think you asked for this?

fjmolinas commented on sys/trace/trace.c:
```diff
@@ -0,0 +1,63 @@
+#define TRACE_BUFSIZE 512
+#endif
+
+typedef struct {
```
Can you uncrustify?

kaspar030 replied:
done

fjmolinas commented on sys/trace/trace.c:
```diff
@@ -0,0 +1,63 @@
+
+#ifndef TRACE_BUFSIZE
+#define TRACE_BUFSIZE 512
+#endif
```
Why not have this in the header?

kaspar030 replied:
because it is not part of the interface. should it be?

fjmolinas replied:
Don't know, so lets leave it that way for now.

fjmolinas commented on sys/trace/trace.c:
```diff
@@ -0,0 +1,63 @@
+{
+    unsigned state = irq_disable();
+    tracebuf[tracebuf_pos % TRACE_BUFSIZE] = (tracebuf_entry_t){ .time = xtimer_now_usec(), .val = val };
+    tracebuf_pos++;
```
Should the value be clamped? Is a modulo operation expensive?

fjmolinas replied:
If so maybe add a mention in the documentation. if someone was tracing in a loop for example it could eventually get overrun depending on the arch.

kaspar030 replied:
yeah, modulo is on the expensive side, unless there's an fpu. also quite large if it has to import software fp support. what do you mean with 'overrun'?

kaspar030 replied:
> unless there's an fpu.

... or the value is a power of two.

fjmolinas replied:
Bad choice of words, I mean that it will eventually start overriding buffer content if not `dumped` or `reset`. Just documenting this in the header should be enough.

aabadie commented:
> Should I try to add the callsite to the log? It might be possible using '__builtin_return_address()', but would require the noinline attribute. It could be optional if performance is hurt a lot.

I don't have any strong opinion on this. The developer is already supposed to know where he added calls to trace. Do you think it's worth ?

fjmolinas commented:
Needs rebase now!

fjmolinas commented:
@kaspar030 seems like this one just needs some blacklisting.

kaspar030 commented:
> @kaspar030 seems like this one just needs some blacklisting.

done. also fixed the doxygen group, and added a note about the ring buffer behaviour of the buffer. Sorry for the late addition!

fjmolinas commented:
Please squash!

fjmolinas commented:
And squash again pleas :)
